plan/expr_to_pb.go

```diff
@@ -154,147 +151,138 @@ func scalarFuncToPBExpr(client kv.Client, expr *expression.ScalarFunction) (*tip
 	if expr1 == nil {
-		return nil, nil
+		return nil
 	}
 	return &tipb.Expr{
 		Tp:       tp,
-		Children: []*tipb.Expr{expr0, expr1}}, nil
+		Children: []*tipb.Expr{expr0, expr1}}
 }
 
-func inToPBExpr(client kv.Client, expr *expression.ScalarFunction) (*tipb.Expr, error) {
+func inToPBExpr(client kv.Client, expr *expression.ScalarFunction) *tipb.Expr {
 	if !client.SupportRequestType(kv.ReqTypeSelect, int64(tipb.ExprType_In)) {
-		return nil, nil
+		return nil
 	}
 
-	pbExpr, err := exprToPB(client, expr.Args[0])
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
+	pbExpr := exprToPB(client, expr.Args[0])
 	if pbExpr == nil {
-		return nil, nil
-	}
-	listExpr, err := constListToPB(client, expr.Args[1:])
-	if err != nil {
-		return nil, errors.Trace(err)
+		return nil
 	}
+	listExpr := constListToPB(client, expr.Args[1:])
 	if listExpr == nil {
-		return nil, nil
+		return nil
 	}
 	return &tipb.Expr{
 		Tp:       tipb.ExprType_In,
-		Children: []*tipb.Expr{pbExpr, listExpr}}, nil
+		Children: []*tipb.Expr{pbExpr, listExpr}}
 }
 
-func notToPBExpr(client kv.Client, expr *expression.ScalarFunction) (*tipb.Expr, error) {
+func notToPBExpr(client kv.Client, expr *expression.ScalarFunction) *tipb.Expr {
 	if !client.SupportRequestType(kv.ReqTypeSelect, int64(tipb.ExprType_Not)) {
-		return nil, nil
+		return nil
 	}
 
-	child, err := exprToPB(client, expr.Args[0])
-	if err != nil {
-		return nil, errors.Trace(err)
-	}
+	child := exprToPB(client, expr.Args[0])
 	if child == nil {
-		return nil, nil
+		return nil
 	}
 	return &tipb.Expr{
 		Tp:       tipb.ExprType_Not,
-		Children: []*tipb.Expr{child}}, nil
+		Children: []*tipb.Expr{child}}
 }
 
-func constListToPB(client kv.Client, list []expression.Expression) (*tipb.Expr, error) {
+func constListToPB(client kv.Client, list []expression.Expression) *tipb.Expr {
 	if !client.SupportRequestType(kv.ReqTypeSelect, int64(tipb.ExprType_ValueList)) {
-		return nil, nil
+		return nil
 	}
 
 	// Only list of *expression.Constant can be push down.
 	datums := make([]types.Datum, 0, len(list))
 	for _, expr := range list {
 		v, ok := expr.(*expression.Constant)
 		if !ok {
-			return nil, nil
+			return nil
 		}
-		if d, err := datumToPBExpr(client, v.Value); err != nil {
-			return nil, errors.Trace(err)
-		} else if d == nil {
-			return nil, nil
+		d := datumToPBExpr(client, v.Value)
+		if d == nil {
+			return nil
 		}
 		datums = append(datums, v.Value)
 	}
 	return datumsToValueList(datums)
 }
 
-func datumsToValueList(datums []types.Datum) (*tipb.Expr, error) {
+func datumsToValueList(datums []types.Datum) *tipb.Expr {
 	// Don't push value list that has different datum kind.
 	prevKind := types.KindNull
 	for _, d := range datums {
 		if prevKind == types.KindNull {
 			prevKind = d.Kind()
 		}
 		if !d.IsNull() && d.Kind() != prevKind {
-			return nil, nil
+			return nil
 		}
 	}
 	err := types.SortDatums(datums)
 	if err != nil {
-		return nil, errors.Trace(err)
+		log.Error(err.Error())
```

Comment:
I think this error shouldn't make sql fail. Instead, it can forbid this condition to push down.